src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java

```diff
@@ -366,65 +423,94 @@ public void processElement(DoFn<KV<Contig, Iterable<Read>>, String>.ProcessConte
       final Contig shardContig = shard.getKey();
       final Iterable<Read> reads = shard.getValue();
       final boolean isFirstShard = shardContig.equals(headerInfo.firstShard);
-
       if (isFirstShard) {
         LOG.info("Writing first shard " + shardContig);
       } else {
         LOG.info("Writing non-first shard " + shardContig);
       }
 
-      final String writeResult = writeShard(headerInfo.header,
-          shardContig, reads,
-          c.getPipelineOptions().as(ShardedBAMWritingOptions.class),
-          isFirstShard);
-      c.output(writeResult);
+      int numRetriesLeft = MAX_RETRIES_FOR_WRITING_A_SHARD;
+      boolean done = false;
+      do {
+       try {
+        final String writeResult = writeShard(headerInfo.header,
+            shardContig, reads,
+            c.getPipelineOptions().as(ShardedBAMWritingOptions.class),
+            isFirstShard);
+        c.output(writeResult);
+        done = true;
+       } catch (IOException iox) {
+         LOG.warning("Write shard failed for " + shardContig + ": " + iox.getMessage());
```

Comment:
For another PR or file a github issue: consider exponential backoff for the GCS client so that there is a delay between retries.

https://developers.google.com/api-client-library/java/google-http-java-client/backoff?hl=en

Here's how we use it for Genomics v1beta2 https://github.com/googlegenomics/utils-java/blob/ad114c01ab81db136f3b0a59f1dcdcbfad94662b/src/main/java/com/google/cloud/genomics/utils/GenomicsFactory.java#L350

gRPC has it built in: https://github.com/grpc/grpc-java/blob/a508c1d4f5024c2a5fd6d103edcab43542293ae7/core/src/main/java/io/grpc/ChannelImpl.java#L75